### PR TITLE
nanocoap: ignore replies with wrong message ID

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This modifies the nanocoap sock request code to ignore potential replies with wrong message ID's.

### Testing procedure

I tested this with `tests/nanocoap_cli`. The following patch will ensure that nanocoap always responds with incorrect ID's and should trigger:

```Patch
diff --git a/sys/net/application_layer/nanocoap/nanocoap.c b/sys/net/application_layer/nanocoap/nanocoap.c
index 2c23f4d3f1..cdea2fcf3b 100644
--- a/sys/net/application_layer/nanocoap/nanocoap.c
+++ b/sys/net/application_layer/nanocoap/nanocoap.c
@@ -469,7 +469,7 @@ ssize_t coap_build_reply(coap_pkt_t *pkt, unsigned code,
     }
 
     coap_build_hdr((coap_hdr_t *)rbuf, type, pkt->token, tkl, code,
-                   ntohs(pkt->hdr->id));
+                   ntohs(pkt->hdr->id) + 1);
     coap_hdr_set_type((coap_hdr_t *)rbuf, type);
     coap_hdr_set_code((coap_hdr_t *)rbuf, code);
```

### Issues/PRs references

Stolen from the SUIT coap code

### CODE REVIEW

File: sys/net/application_layer/nanocoap/sock.c [hunk 1]
```diff
@@ -78,6 +108,11 @@ ssize_t nanocoap_request(coap_pkt_t *pkt, sock_udp_ep_t *local, sock_udp_ep_t *r
                 DEBUG("nanocoap: error parsing packet\n");
                 res = -EBADMSG;
             }
+            else if (coap_get_id(pkt) != id) {
+                res = -EBADMSG;
+                continue;
+            }
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c]: Does it make sense not to decrement the retry-count if you got a 'bad' message?

File: sys/net/application_layer/nanocoap/sock.c [hunk 2]
```diff
@@ -46,29 +64,41 @@ ssize_t nanocoap_request(coap_pkt_t *pkt, sock_udp_ep_t *local, sock_udp_ep_t *r
...
+        if (res == -EAGAIN) {
+            res = sock_udp_send(&sock, buf, pdu_len, NULL);
+            if (res <= 0) {
+                DEBUG("nanocoap: error sending coap request, %d\n", (int)res);
+                break;
+            }
         }
 
-        res = sock_udp_recv(&sock, buf, len, timeout, NULL);
+        res = sock_udp_recv(&sock, buf, len, deadline_left(deadline), NULL);
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c]: I reckon this is to account for the time 'lost' by sending the UDP packet?

File: sys/net/application_layer/nanocoap/sock.c (hunk 1 above)
Comment (bergzand) [reply sys/net/application_layer/nanocoap/sock.c]: Yes…

File: sys/net/application_layer/nanocoap/sock.c (hunk 2 above)
Comment (bergzand) [reply sys/net/application_layer/nanocoap/sock.c]: This is to ensure that the total time listened is equal to the timeout, even when a recv call returned a packet with incorrect message ID.

Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c]: Better add a comment about it then.

File: sys/net/application_layer/nanocoap/sock.c (hunk 1 above)
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c]: Oh I see you only re-transmit the message if `res == -EAGAIN` and you just set it to `-EBADMSG` here.

File: sys/net/application_layer/nanocoap/sock.c [hunk 3]
```diff
@@ -78,6 +108,11 @@ ssize_t nanocoap_request(coap_pkt_t *pkt, sock_udp_ep_t *local, sock_udp_ep_t *r
                 DEBUG("nanocoap: error parsing packet\n");
                 res = -EBADMSG;
             }
+            else if (coap_get_id(pkt) != id) {
+                res = -EBADMSG;
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c]:
```suggestion
                /* don't send the request again */
                res = -EBADMSG;
```

File: sys/net/application_layer/nanocoap/sock.c (hunk 2 above)
Comment (bergzand) [reply sys/net/application_layer/nanocoap/sock.c]: Added

File: sys/net/application_layer/nanocoap/sock.c (hunk 3 above)
Comment (bergzand) [reply sys/net/application_layer/nanocoap/sock.c]: Added

**Comment by bergzand** [comment]
After a short moment of clarity I noticed that this PR is broken. The content of the reply replaces the buffer content of the request. If a response with a bad message ID lands but no correct reply is received within the timeout, the next retransmission will transmit the bad message.

File: sys/net/application_layer/nanocoap/sock.c [hunk 4]
```diff
@@ -30,11 +30,29 @@
...
 
+static inline uint32_t deadline_from_interval(int32_t interval)
+{
+    assert(interval >= 0);
+    return xtimer_now_usec() + (uint32_t)interval;
+}
+
+static inline uint32_t deadline_left(uint32_t deadline)
+{
+    int32_t left = (int32_t)(deadline - xtimer_now_usec());
```
Comment (bergzand) [inline sys/net/application_layer/nanocoap/sock.c]: This is incorrect :cry:

**Comment by fjmolinas** [comment]
> After a short moment of clarity I noticed that this PR is broken. The content of the reply replaces the buffer content of the request. If a response with a bad message ID lands but no correct reply is received within the timeout, the next retransmission will transmit the bad message.

Is there an equivalent of a resend buffer in nanocoap? I know GCOAP uses one for this `CONFIG_GCOAP_RESEND_BUFS_MAX`?

**Comment by miri64** [comment]
Ping @bergzand?

File: sys/net/application_layer/nanocoap/sock.c (hunk 4 above)
Comment (fjmolinas) [reply sys/net/application_layer/nanocoap/sock.c:41]: hm?

**Comment by fjmolinas** [comment]
Any updates on this @bergzand?

**Comment by bergzand** [comment]
I still have to figure out a way to work around [this](https://github.com/RIOT-OS/RIOT/pull/13707#issuecomment-606465166) issue

**Comment by aabadie** [comment]
This seems to fix a bug, so reopen and the label (which will disable stalebot)

**Comment by benpicco** [comment]
Isn't this already [fixed in master](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/application_layer/nanocoap/sock.c#L144)?

**Comment by fjmolinas** [comment]
> Isn't this already [fixed in master](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/application_layer/nanocoap/sock.c#L144)?

yep seems to be fixed, and most of this PR does not apply anymore

**Comment by benpicco** [comment]
Fixed in master

**Pull request closed without merging**